PWGHF/TableProducer/candidateSelectorB0ToDPi.cxx

```diff
@@ -121,40 +125,79 @@ struct HfCandidateSelectorB0ToDPi {
+  /// Apply PID selection
+  /// \param pidTrackPi is the PID status of trackPi (prong1 of B0 candidate)
+  /// \return true if prong1 of B0 candidate passes all selections
+  template <typename T = int>
+  bool selectionPID(const T& pidTrackPi)
   {
-    for (auto const& hfCandB0 : hfCandB0s) { // looping over B0 candidates
+    if (!acceptPIDNotApplicable && pidTrackPi != TrackSelectorPID::Status::PIDAccepted) {
+      return false;
+    }
+    if (acceptPIDNotApplicable && pidTrackPi == TrackSelectorPID::Status::PIDRejected) {
+      return false;
+    }
 
-      int statusB0 = 0;
+    return true;
+  }
+
+  using TracksPIDWithSel = soa::Join<aod::BigTracksPIDExtended, aod::TrackSelection>;
 
-      // check if flagged as B0 → D- π+
+  void process(aod::HfCandB0 const& hfCandsB0, soa::Join<aod::HfCand3Prong, aod::HfSelDplusToPiKPi> const&, TracksPIDWithSel const&)
+  {
+    int statusB0ToDPi = 0;
+
+    TrackSelectorPID selectorPion(kPiPlus);
+    selectorPion.setRangePtTPC(ptPidTpcMin, ptPidTpcMax);
+    selectorPion.setRangeNSigmaTPC(-nSigmaTpcMax, nSigmaTpcMax);
+    selectorPion.setRangeNSigmaTPCCondTOF(-nSigmaTpcCombinedMax, nSigmaTpcCombinedMax);
+    selectorPion.setRangePtTOF(ptPidTofMin, ptPidTofMax);
+    selectorPion.setRangeNSigmaTOF(-nSigmaTofMax, nSigmaTofMax);
+    selectorPion.setRangeNSigmaTOFCondTPC(-nSigmaTofCombinedMax, nSigmaTofCombinedMax);
+
+    for (const auto& hfCandB0 : hfCandsB0) {
+      // check if flagged as B0 → D π
       if (!TESTBIT(hfCandB0.hfflag(), hf_cand_b0::DecayType::B0ToDPi)) {
-        hfSelB0ToDPiCandidate(statusB0);
-        // Printf("B0 candidate selection failed at hfflag check");
+        hfSelB0ToDPiCandidate(statusB0ToDPi);
+        // LOGF(info, "B0 candidate selection failed at hfflag check");
         continue;
       }
+      SETBIT(statusB0ToDPi, aod::SelectionStep::RecoSkims); // RecoSkims = 0 --> statusB0ToDPi = 1
 
-      // D is always index0 and pi is index1 by default
-      // auto candD = hfCandD.prong0();
       auto candD = hfCandB0.prong0_as<soa::Join<aod::HfCand3Prong, aod::HfSelDplusToPiKPi>>();
-      auto trackPi = hfCandB0.prong1_as<aod::BigTracksPID>();
+      auto trackPi = hfCandB0.prong1_as<TracksPIDWithSel>();
 
       // topological cuts
       if (!selectionTopol(hfCandB0, candD, trackPi)) {
-        hfSelB0ToDPiCandidate(statusB0);
-        // Printf("B0 candidate selection failed at selection topology");
+        hfSelB0ToDPiCandidate(statusB0ToDPi);
+        // LOGF(info, "B0 candidate selection failed at topology selection");
         continue;
       }
+      SETBIT(statusB0ToDPi, aod::SelectionStep::RecoTopol); // RecoTopol = 1 --> statusB0ToDPi = 3
+
+      // track-level PID selection
+      if (usePid) {
+        if (!TESTBIT(candD.isSelDplusToPiKPi(), aod::SelectionStep::RecoPID)) { // safety
+          LOG(warning) << "PID not enabled in DplusToPiKPi selector. Set selectionFlagD=7 in hf-candidate-creator-b0";
```

Comment:
I don't understand this part. The condition tests whether the D+ candidate passed the PID selection. Why does the warning mention enabling PID in the D+ selector and adjusting the selection flag in the B0 candidate creator?

Reply:
OK, I think I see what you mean now. You want to sync the application of PID between the B0 candidate creator (`selectionFlagD`) and the B0 candidate selector (`usePid`), but it has nothing to do with the PID selection in the D+ selector (which cannot be enabled or disabled).
If you run both B0 workflows (creator and selector), you could inspect the value of `selectionFlagD` from the selector and compare it with `usePid`. Of course this only works if you run both workflows.
Otherwise, one has to either apply both selections in the same workflow or just cross-check the settings.

Reply:
In any case, the warning message should be rephrased and not refer to the D+ selector.

Reply:
Yes, I agree. I had a poor choice of words. Indeed, whatever is the value of `selectionFlagD`, the PID selection is applied in Dplus selector, `selectionFlagD` only affects the `Filter` of Dplus candidates in the B0 creator.

So, do you think we should keep this approach or simply for for option 2 (where there is no more conflict between `selectionFlagD` and `usePid`)?

Reply:
The values in `HFSELDPLUS` should definitely be used and I guess the point of using it in the B0 creator instead of in the selector is to reduce the number of candidates stored in the table and processed in the next steps. All that makes sense. Therefore, if you want to sync the settings of `selectionFlagD` and `usePid` (instead of keeping them independent), I would be in favour of adding the value inspection. I would suggest putting it all inside `init`, including the warning message, such that it is checked (and reported in the log) only once instead of for each candidate.

Reply:
Thanks for the reply @vkucera . But I don't know the syntax to get the value of the configurable `selectionFlagD` (defined in `candidateCreatorB0.cxx`) inside the `init` function of `candidateSelectorB0ToDPi.cxx` (I did not find it in the documentation).

Reply:
Can you try to take inspiration from this snippet? https://github.com/AliceO2Group/O2Physics/commit/80b870054a0480d1b01e70e97532b0db030975ce#diff-f39bae092160ecb233c90fe7d2e546e3d12b169c54dddc3db216aa27a400ea95L582-L587

Reply:
I am pretty sure because I changed the value of `selectionFlagD` in my `.json` and it still displayed the hardcoded "default" value.

Reply:
Hi @iouribelikov , thanks for the link. I didn't know. That is unfortunate. In that case, we cannot guarantee that the two configurations are synced.
@AlexBigO how do you want to detect the cases `usePid=false` and `selectionFlagD=7`?

Reply:
Indeed, as a temporary solution, it might work. Please add a `FIXME` comment with the link to the JIRA ticket at the relevant place in the B0 selector.

Reply:
In which function do you fill the table with `myflag`? I suppose that even filling in `init` should work, since there the actual values are used for sure.

Reply:
- in `candidateCreatorB0.cxx`
First, I made a mistake by defining `int myflag` before `init` or `process`. If I define it inside `init` or `process` I get the configurable value set in `.json`. So far so good.
Now, if I try filling my table in `init`, `hf-candidate-creator-b0` crashes during the execution
But it works if I fill the table in `process` (I put a condition so it only gets filled  only once).
- in `candidateSelectorB0ToDPi.cxx`
If I try to subscribe to my AOD table in `init` (on top of `InitContext const&`), O2Physics does not compile. But, if I subscribe to the table in `process` (as it is usually done), then it works.
- Recap
I fill and read my table in both `process` functions of B0 creator and selector, respectively. Is that acceptable as a temporary solution? Or is there a way to do it via `init`?

Reply:
Thanks for the tests. I think what you describe in the recap is acceptable for now.